File: src/main/java/io/github/hhy/bookmark/manager/MyBookmarkManager.java
package io.github.hhy.bookmark.manager;

import com.intellij.openapi.application.ApplicationInfo;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.BuildNumber;
import io.github.hhy.bookmark.element.Element;

import java.util.List;

public interface MyBookmarkManager {

    static MyBookmarkManager getBookmarkManager() {
        BuildNumber build = ApplicationInfo.getInstance().getBuild();
        if (build.getBaselineVersion() >= 231) {
            return new HighVersionBookManager();
        }
        return new LowVersionBookManager();
    }

    /**
     * get all
     *
     * @return
     */
    public List<Element> getAllBookmarks(Project project);

    /**
     * Add
     * @param project
     * @param elements
     */
    public void addBookmarks(Project project, List<Element> elements);
}

File: src/main/java/io/github/hhy/bookmark/manager/MyBookmarkManager.kt
package io.github.hhy.bookmark.manager

import com.intellij.ide.bookmark.Bookmark
import com.intellij.ide.bookmark.BookmarkGroup
import com.intellij.openapi.project.Project
import io.github.hhy.bookmark.element.BookmarkElement
import io.github.hhy.bookmark.element.GroupElement

interface MyBookmarkManager {

    /**
     * get all
     *
     * @return
     */
    fun getAllBookmarks(project: Project): Map<String /* groupName */, Map<String /* bookmarkElementKey */, BookmarkElement>>

    /**
     * Add
     *
     * @param project
     * @param elements
     */
    fun addGroup(project: Project, ele: GroupElement): BookmarkGroup

    /**
     *
     */
    fun addBookmark(project: Project, groupName :String, ele: BookmarkElement)

    /**
     * 删除无效的书签
     *
     * @param project
     */
    fun removeInvalid(project: Project): List<Bookmark> {
        return emptyList()
    }

    companion object {
        val bookmarkManager: MyBookmarkManager = HighVersionBookManager()
    }
}
